File: tests/gpd_test.cc
//
// APFEL++ 2017
//
// Author: Valerio Bertone: valerio.bertone@cern.ch
//

#include <apfel/apfelxx.h>

#include <functional>

int main()
{
  // x-space grid
  const apfel::Grid g{{apfel::SubGrid{100,1e-5,3}, apfel::SubGrid{60,1e-1,3}, apfel::SubGrid{50,6e-1,3}, apfel::SubGrid{50,8e-1,3}}};

  // Initial scale
  const double mu0 = sqrt(2);

  // Vectors of masses and thresholds
  const std::vector<double> Thresholds = {0, 0, 0, sqrt(2), 4.5, 175};;

  // Perturbative order
  const int PerturbativeOrder = 0;

  // Skewness
  const double xi = 0.01;

  // Running coupling
  const double AlphaQCDRef = 0.35;
  const double MuAlphaQCDRef = sqrt(2);
  apfel::AlphaQCD a{AlphaQCDRef, MuAlphaQCDRef, Thresholds, PerturbativeOrder};
  const apfel::TabulateObject<double> Alphas{a, 100, 0.9, 1001, 3};
  const auto as = [&] (double const& mu) -> double{ return Alphas.Evaluate(mu); };

  // Initialize GPD evolution objects
  const auto GpdObj = InitializeGpdObjects(g, Thresholds, xi);

  // Construct the DGLAP objects
  const auto EvolvedGPDs = BuildDglap(GpdObj,   apfel::LHToyPDFs, mu0, PerturbativeOrder, as);

  // Tabulate PDFs
  const apfel::TabulateObject<apfel::Set<apfel::Distribution>> TabulatedGPDs{*EvolvedGPDs, 50, 1, 1000, 3};

  // Final scale
  const double mu = 10;

  // Compute results
  std::cout << std::scientific << "Direct evolution (4th order Runge-Kutta) from Q0 = " << mu0 << " GeV to Q = " << mu << " GeV... ";

  // Evolve GPDs to the final Scale
  apfel::Timer t;
  const std::map<int, apfel::Distribution> gpds = apfel::QCDEvToPhys(EvolvedGPDs->Evaluate(mu).GetObjects());
  t.stop();

  // Print results
  const std::vector<double> xlha = {1e-3, 1e-2, 1e-1, 3e-1, 5e-1, 7e-1, 9e-1};

  std::cout << "\nAlphaQCD(Q) = " << Alphas.Evaluate(mu) << std::endl;
  std::cout << "\n   x    "
            << "   u-ubar   "
            << "   d-dbar   "
            << " 2(ubr+dbr) "
            << "   c+cbar   "
            << "    gluon   "
            << std::endl;

  std::cout << "Direct Evolution:" << std::endl;
  for (auto const& x : xlha)
    {
      std::cout.precision(1);
      std::cout << x;
      std::cout.precision(4);
      std::cout << "  " << (gpds.at(2) - gpds.at(-2)).Evaluate(x)
                << "  " << (gpds.at(1) - gpds.at(-1)).Evaluate(x)
                << "  " << 2 * (gpds.at(-2) + gpds.at(-1)).Evaluate(x)
                << "  " << (gpds.at(4) + gpds.at(-4)).Evaluate(x)
                << "  " << gpds.at(0).Evaluate(x)
                << std::endl;
    }
  std::cout << "\n";

  return 0;
}

File: pywrap/gpd_test.py
import apfelpy as ap
import numpy as np

# x-space grid
g = ap.Grid([ap.SubGrid(100,1e-5,3), ap.SubGrid(200,1e-1,3), ap.SubGrid(100,9e-1,3)])
#g = ap.Grid([ap.SubGrid(30,1e-5,3), ap.SubGrid(20,1e-1,3), ap.SubGrid(10,9e-1,3)])

# Initial scale
mu0 = np.sqrt(2)

# Vectors of masses and thresholds
Thresholds = [0, 0, 0, np.sqrt(2), 4.5, 175]

# Perturbative order
PerturbativeOrder = 0

# Skewness
xi = 0.3

# Running coupling
a = ap.AlphaQCD(0.35, np.sqrt(2), Thresholds, PerturbativeOrder)
Alphas = ap.TabulateObject(a, 100, 0.9, 1001, 3)

# Initialize QCD evolution objects
GpdObj   = ap.initializers.InitializeGpdObjects(g, Thresholds, xi)
#GpdObjOp = ap.initializers.InitializeGpdObjects(g, Thresholds, xi, True)

# Construct the DGLAP objects
EvolvedGPDs = ap.builders.BuildDglap(GpdObj,   ap.utilities.LHToyPDFs, mu0, PerturbativeOrder, Alphas.Evaluate)
#EvolvedOps  = ap.builders.BuildDglap(GpdObjOp,                         mu0, PerturbativeOrder, Alphas.Evaluate)

# Tabulate GPDs
TabulatedGPDs = ap.TabulateObjectSetD(EvolvedGPDs, 30, 1, 100, 3)

# Tabulate Operators
#TabulatedOps = ap.TabulateObjectSetO(EvolvedOps, 30, 1, 100, 3)

# Final scale
mu = 10

# Compute results
print("Direct evolution (4th order Runge-Kutta) from Q0 = " , mu0 , " GeV to Q = " , mu , " GeV... ")

# Evolve GPDs to the final Scale
t = ap.Timer()
gpds = ap.utilities.QCDEvToPhys(EvolvedGPDs.Evaluate(mu).GetObjects())
t.stop()

print("Interpolation of the tabulated GPDs... ")
t.start()
tgpds = ap.utilities.QCDEvToPhys(TabulatedGPDs.Evaluate(mu).GetObjects())
t.stop()

#print("Interpolation of the tabulated evolution operators... ")
#t.start()
#tops = TabulatedOps.Evaluate(mu)
#t.stop()

# Print results
xlha = [1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 3e-1, 5e-1, 7e-1, 9e-1]
print("\nAlphaQCD(Q) = " , Alphas.Evaluate(mu))
print("\n   x    "
          , "   u-ubar   "
          , "   d-dbar   "
          , " 2(ubr+dbr) "
          , "   c+cbar   "
          , "    gluon   "
         )

#print("Direct Evolution:")
#for x in xlha:
#    print(format(x, ".1e"), " ",
#              format((gpds[2] - gpds[-2]).Evaluate(x), ".4e"), " ",
#              format((gpds[1] - gpds[-1]).Evaluate(x), ".4e"), " ",
#              format(2 * (gpds[-2] + gpds[-1]).Evaluate(x), ".4e"), " ",
#              format((gpds[4] + gpds[-4]).Evaluate(x), ".4e"), " ",
#              format(gpds[0].Evaluate(x), ".4e"))

print("\nInterpolation on the GPD table (all x for each Q):")
for x in xlha:
    print(format(x, ".1e"), " ",
              format((tgpds[2] - tgpds[-2]).Evaluate(x), ".4e"), " ",
              format((tgpds[1] - tgpds[-1]).Evaluate(x), ".4e"), " ",
              format(2 * (tgpds[-2] + tgpds[-1]).Evaluate(x), ".4e"), " ",
              format((tgpds[4] + tgpds[-4]).Evaluate(x), ".4e"), " ",
              format(tgpds[0].Evaluate(x), ".4e"))

print("\nInterpolation on the GPD table as a map (x and Q independently):")
for x in xlha:
    DistMap = ap.utilities.QCDEvToPhys(TabulatedGPDs.EvaluateMapxQ(x,mu))
    print(format(x, ".1e"), " ",
              format(DistMap[2] - DistMap[-2], ".4e"), " ",
              format(DistMap[1] - DistMap[-1], ".4e"), " ",
              format(2 * (DistMap[-2] + DistMap[-1]), ".4e"), " ",
              format(DistMap[4] + DistMap[-4], ".4e"), " ",
              format(DistMap[0], ".4e"))

t.stop()
